README.rst
A collection of Python library modules
======================================

This package provides a collection of small helper functions and
classes.


System requirements
-------------------

Python:

+ Python 3.4 or newer.

Required library packages:

+ None

Optional library packages:

+ `setuptools_scm`_

  The version number is managed using this package.  All source
  distributions add a static text file with the version number and
  fall back using that if `setuptools_scm` is not available.  So this
  package is only needed to build out of the plain development source
  tree as cloned from GitHub.

+ `pytest`_ >= 3.0

  Only needed to run the test suite.

+ `distutils-pytest`_

  Only needed to run the test suite.


Copyright and License
---------------------

- Copyright 2015–2019
  Helmholtz-Zentrum Berlin für Materialien und Energie GmbH
- Copyright 2019–2020 Rolf Krahl

Licensed under the Apache License, Version 2.0 (the "License"); you
may not use this file except in compliance with the License.  You may
obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
implied.  See the License for the specific language governing
permissions and limitations under the License.


.. _setuptools_scm: https://github.com/pypa/setuptools_scm/
.. _pytest: https://pytest.org/
.. _distutils-pytest: https://github.com/RKrahl/distutils-pytest


CHANGES.rst
History of changes to rklib
===========================

0.1.dev (not yet released)
    + Initial release.  The modules in this package have been taken
      from other packages.  Some of them I have written for my job at
      Helmholtz-Zentrum Berlin für Materialien und Energie (HZB),
      others stem from private projects of mine:

      - :mod:`rklib.filelock`: a first version of this code has been
	written for an internal project at HZB in February 2015.  A
	later version has been reused in `tiny-init`_ in July 2018.

      - :mod:`rklib.units`: the classes in this module have first been
	written for `icat-server-test`_ in January 2017.

.. _tiny-init: https://github.com/RKrahl/tiny-init
.. _icat-server-test: https://github.com/RKrahl/icat-server-test


tests/test_00.py
"""Dummy test.
"""

import pytest

def test_import():
    """Try importing the module.
    """
    import rklib
    print(rklib)


tests/test_01_units.py
"""Test module rklib.units.
"""

import pytest
from rklib.units import Time, MemorySize


def test_time():
    """Test class Time.
    """
    t1 = Time(2)
    assert str(t1) == "2.000 s"
    assert Time("2s") == t1
    t2 = Time(2.0)
    assert str(t2) == "2.000 s"
    assert Time("2.0 s") == t2
    t3 = Time(90*60)
    assert str(t3) == "1.500 h"
    assert Time("1.50 h") == t3
    t4 = Time(30*60*60)
    assert str(t4) == "1.250 d"
    assert Time("1.25 d") == t4
    t5 = Time(0.0205)
    assert str(t5) == "20.500 ms"
    assert Time("20.50ms") == t5

    t6 = 2*Time(1)
    assert str(t6) == "2.000 s"
    t7 = 1.5*Time("1h")
    assert str(t7) == "1.500 h"
    assert t7 == 5400.0
    t8 = Time("2d") + 4*(Time("1h") + Time("12min"))
    assert str(t8) == "2.200 d"
    t9 = Time("1h") + Time("15 min") - 0.5*Time("1h")
    assert str(t9) == "45.000 min"


def test_memsize():
    """Test class MemorySize.
    """
    s1 = MemorySize(2)
    assert str(s1) == "2 B"
    assert MemorySize("2B") == s1
    s2 = MemorySize(2.0)
    assert str(s2) == "2 B"
    assert MemorySize("2.0 B") == s2
    s3 = MemorySize(1536*1024)
    assert str(s3) == "1.50 MiB"
    assert MemorySize("1.50 MiB") == s3
    s4 = MemorySize(30.12*1024*1024*1024*1024)
    assert str(s4) == "30.12 TiB"
    assert MemorySize("30.12 TiB") == s4

    s5 = 2*MemorySize(1)
    assert str(s5) == "2 B"
    s6 = MemorySize("2 GiB") + 5*MemorySize("256 MiB")
    assert str(s6) == "3.25 GiB"
    assert s6 == 3328*1024*1024
    s7 = MemorySize("1 PiB") + 3*MemorySize("128 TiB") - MemorySize("512 TiB")
    assert str(s7) == "896.00 TiB"


init.py
#! python3

import argparse
from pathlib import Path
import string
import subprocess


argparser = argparse.ArgumentParser(description="Initialize the repository.")
argparser.add_argument("distname", help="name of the package")
args = argparser.parse_args()


tagmsg = """Tag initial commit.  This is not a release version.

Add this tag solely to make sure there is a tag somewhere in the
history that looks like a version number.
"""
subprocess.check_call(["git", "tag", "-a", "-m", tagmsg, "0.0"])


distname_files = (
    Path(".gitignore"),
    Path("Makefile"),
    Path("python-skel.spec"),
    Path("setup.py"),
    Path("tests", "test_00.py"),
)

for path in distname_files:
    with path.open("rt") as f:
        s = string.Template(f.read())
    with path.open("wt") as f:
        f.write(s.safe_substitute(distname=args.distname))
    subprocess.check_call(["git", "add", str(path)])

subprocess.check_call(["git", "mv",
                       "python-skel.spec", "python-%s.spec" % args.distname])
subprocess.check_call(["git", "rm", "init.py"])
subprocess.check_call(["git", "commit", "-m", "Set the name of the package."])

print("""Name of the package set.

Next steps: fix the url in setup.py and adapt the description of the
package in the doc string in setup.py and in the README.rst.""")


rklib/file.py
"""Tools for reading and writing files.
"""

import hashlib
from pathlib import Path
import zlib


class _CRC32():
    """Provide an hashlib compatible interface to zlib.crc32().
    """
    def __init__(self):
        self.crc32 = 0
    def update(self, chunk):
        self.crc32 = zlib.crc32(chunk, self.crc32)
    def hexdigest(self):
        return "%x" % (self.crc32 & 0xffffffff)

def _new_checksum(name):
    if name == "crc32":
        return _CRC32()
    else:
        return hashlib.new(name)

def copy_file_crc(src_file, dest_file, checksums=None, chunksize=8192):
    """Copy a file and calculate checksums on the fly.
    """

    if not hasattr(src_file, "read"):
        if not hasattr(src_file, "open"):
            src_file = Path(src_file)
        with src_file.open("rb") as sf:
            return copy_file_crc(sf, dest_file, checksums, chunksize)

    if not hasattr(dest_file, "write"):
        if not hasattr(dest_file, "open"):
            dest_file = Path(dest_file)
        with dest_file.open("wb") as df:
            return copy_file_crc(src_file, df, checksums, chunksize)

    checksums = checksums or []
    m = { c:_new_checksum(c) for c in checksums }
    while True:
        chunk = src_file.read(chunksize)
        if not chunk:
            break
        for c in checksums:
            m[c].update(chunk)
        dest_file.write(chunk)
    return tuple( m[c].hexdigest() for c in checksums )


scripts/filelock.py
#! python
"""Try acquiring a fcntl file lock on a file.

The lock may either be shared or exclusive.  Run a command while
holding the lock.
"""

import argparse
import fcntl
import logging
import os
from pathlib import Path
import shlex
import subprocess
import sys

from rklib.filelock import AlreadyLockedError, filelock

try:
    shell = os.environ['SHELL']
except KeyError:
    shell = '/bin/sh'

modes = {'EX':fcntl.LOCK_EX, 'SH':fcntl.LOCK_SH}

cli = argparse.ArgumentParser(description=__doc__.split("\n")[0].rstrip("."))
cli.add_argument("--debug",
                 action='store_const', dest='loglevel',
                 const=logging.DEBUG, default=logging.INFO,
                 help="enable debug output")
cli.add_argument("--read-only", "--ro", action='store_true',
                 help="open the lock file read only")
cli.add_argument("--mode", choices=modes.keys(), default='EX',
                 help="lock mode, either exclusive or shared")
cli.add_argument("-e", "--cmd", default=shell,
                 help="external command to execute while holding the lock")
cli.add_argument('lockfile', type=Path,
                 help="the file to acquire the lock on")
args = cli.parse_args()

prog_name = Path(sys.argv[0]).name if sys.argv[0] else "-"
logformat = "%(asctime)-15s %(name)s %(levelname)s: %(message)s"
logging.basicConfig(level=args.loglevel, format=logformat)
log = logging.getLogger(prog_name)


try:
    with filelock(str(args.lockfile), mode=modes[args.mode], ro=args.read_only):
        log.debug("run %s", args.cmd)
        subprocess.run(shlex.split(args.cmd), cwd=args.lockfile.parent)
except AlreadyLockedError:
    log.critical("%s is already locked", args.lockfile)
    sys.exit(1)


rklib/units.py
"""Human readable quantities of some units.

This modules provides subclasses of :class:`int` or :class:`float` to
denote particular quantities such as time intervalls or amounts of
memory.  The classes provide string representations in a more human
readable form and constructors that accept these string
representations as input.
"""

import re


class Time(float):
    """Human readable time intervals.

    The value is a :class:`float` that indicates the time in seconds.

    >>> t1 = Time(4500)
    >>> str(t1)
    '1.250 h'
    >>> t2 = Time("1.5 d")
    >>> t2
    129600.0
    >>> t3 = Time("36 h")
    >>> t3 == t2
    True
    >>> t4 = Time(1/200)
    >>> str(t4)
    '5.000 ms'
    >>> t5 = 0.5*Time("1 h") + 5*Time("15 min")
    >>> str(t5)
    '1.750 h'
    """
    second = 1
    minute = 60*second
    hour = 60*minute
    day = 24*hour
    millisecond = (1/1000)*second
    units = { 'ms':millisecond, 's':second, 'min':minute, 'h':hour, 'd':day, }
    regexp = re.compile(r'^(\d+(?:\.\d+)?)\s*(ms|s|min|h|d)$')

    def __new__(cls, value):
        if isinstance(value, str):
            m = cls.regexp.match(value)
            if not m:
                raise ValueError("Invalid Time string '%s'" % value)
            v = float(m.group(1)) * cls.units[m.group(2)]
            return super().__new__(cls, v)
        else:
            v = float(value)
            if v < 0:
                raise ValueError("Invalid time value %f" % v)
            return super().__new__(cls, v)

    def __str__(self):
        for u in ['d', 'h', 'min', 's']:
            if self >= self.units[u]:
                return "%.3f %s" % (self / self.units[u], u)
        else:
            return "%.3f ms" % (self / self.units['ms'])

    def __add__(self, other):
        v = super().__add__(other)
        if isinstance(other, Time):
            return Time(v)
        else:
            return v

    def __sub__(self, other):
        v = super().__sub__(other)
        if isinstance(other, Time) and v >= 0:
            return Time(v)
        else:
            return v

    def __rmul__(self, other):
        if type(other) in { float, int }:
            return Time(other*float(self))
        else:
            return super().__rmul__(other)


class MemorySize(int):
    """Human readable amounts of memory.

    The value is an :class:`int` that indicates the number of bytes.

    >>> s1 = MemorySize(2)
    >>> str(s1)
    '2 B'
    >>> s2 = MemorySize("2.0 B")
    >>> s2
    2
    >>> s2 == s1
    True
    >>> s3 = MemorySize(33117290228613)
    >>> str(s3)
    '30.12 TiB'
    >>> s4 = MemorySize("1 PiB") + 3*MemorySize("128 TiB")
    >>> str(s4)
    '1.38 PiB'
    >>> s5 = s4 - MemorySize("512 TiB")
    >>> str(s5)
    '896.00 TiB'
    >>> s5
    985162418487296
    """
    sizeB = 1
    sizeKiB = 1024*sizeB
    sizeMiB = 1024*sizeKiB
    sizeGiB = 1024*sizeMiB
    sizeTiB = 1024*sizeGiB
    sizePiB = 1024*sizeTiB
    sizeEiB = 1024*sizePiB
    units = { 'B':sizeB, 'KiB':sizeKiB, 'MiB':sizeMiB, 'GiB':sizeGiB,
              'TiB':sizeTiB, 'PiB':sizePiB, 'EiB':sizeEiB, }
    regexp = re.compile(r'^(\d+(?:\.\d+)?)\s*(B|KiB|MiB|GiB|TiB|PiB|EiB)$')

    def __new__(cls, value):
        if isinstance(value, str):
            m = cls.regexp.match(value)
            if not m:
                raise ValueError("Invalid MemorySize string '%s'" % value)
            v = float(m.group(1)) * cls.units[m.group(2)]
            return super().__new__(cls, v)
        else:
            v = int(value)
            if v < 0:
                raise ValueError("Invalid size value %d" % v)
            return super().__new__(cls, v)

    def __str__(self):
        for u in ['EiB', 'PiB', 'TiB', 'GiB', 'MiB', 'KiB']:
            if self >= self.units[u]:
                return "%.2f %s" % (self / self.units[u], u)
        else:
            return "%d B" % (int(self))

    def __add__(self, other):
        v = super().__add__(other)
        if isinstance(other, MemorySize):
            return MemorySize(v)
        else:
            return v

    def __sub__(self, other):
        v = super().__sub__(other)
        if isinstance(other, MemorySize) and v >= 0:
            return MemorySize(v)
        else:
            return v

    def __rmul__(self, other):
        if type(other) == int:
            return MemorySize(other*int(self))
        else:
            return super().__rmul__(other)
